##########################################
##                                       #
## ChipNChop multitask ChIP-seq analyzer #
##                                       #
##########################################
##
## Developed by: Gonzalez-Toro, A., Tamargo-Azpilicueta, J., Vazquez-Pacheco, J.
## January, 2021
##

print("", quote=F)
print("Accesing R script. This script allows you to determine the target genes of a transcription factor of interest from narrowPeak file generated by MaCS2.", quote=F)
print("", quote=F)

## Setting working directory to the current file path

args = commandArgs(trailingOnly=T)
setwd(args[1])


print("", quote = F)
print(paste(c("Current working directory is:", getwd())), quote=F)
print("", quote = F)


## Dependencies installation

print("Next libraries are being loaded:", quote=F)

print("ChiPseeker")
library(ChIPseeker)

print("clusterProfiler")
library(clusterProfiler)

print("A. thaliana TxDb transcript metadata")
library(TxDb.Athaliana.BioMart.plantsmart28)

print("A. thaliana annotation")
library(org.At.tair.db)

print("DOSE")
library(DOSE)

print("enrichplot")
library(enrichplot)

print("clusterProfiler")
library(clusterProfiler)

txdb <- TxDb.Athaliana.BioMart.plantsmart28
annotation_atha<-org.At.tair.db

## Reading broad peak files

print("", quote = F)
print("Reading broad peak files from MaCS2 analysis", quote = F)
print("", quote = F)

peaks <- readPeakFile(peakfile = paste0(c(args[2],"_peaks.broadPeak"), collapse = ""),header=FALSE)

covplot(peaks, weightCol = "V5", title = paste0(c("ChIP Peaks of ",args[2], "histone/epigenetic mark over Chromosomes"), collapse = ""))

## Defining the region that is considered a promoter nearby the transcription start site (TSS)

promoter <- getPromoters(TxDb=txdb, upstream=1000, downstream=1000)

## Peak annotation 

print("", quote = F)
print("ChIPseeker is now annotating peaks", quote = F)
print("", quote = F)

peakAnno <- annotatePeak(peak = peaks, tssRegion=c(-1000, 1000), TxDb=txdb)

## Representation of annotated peaks in pie chart

plotAnnoPie(peakAnno,ndigit=0.001)

## Alternatively, by bar plot
## plotAnnoBar(x = peakAnno)

print("", quote = F)
print("Distribution of genomic loci is being displayed. Do check that genes are no further than 1 kb from the estimated TSS. If so, try to manually change parameters of getPromoters and annotatePeak.", quote = F)
print("", quote = F)

plotDistToTSS(peakAnno, title="Distribution of genomic loci relative to TSS", ylab = "Genomic Loci (%) (5' -> 3')")


### Average Profile of ChIP peaks binding to TSS region

tagMatrix <- list(getTagMatrix(peak = peaks, windows=promoter))

plotAvgProf(tagMatrix[[1]], xlim=c(-1000, 1000), resample=100, conf = 0.95)

## Heatmap of ChIP binding to TSS regions

## tagHeatmap is quite bugged. Once available, they might be enabled.

## tagHeatmap(tagMatrix, xlim=c(-1000, 1000), color="red", title="Genomic occupancy of transcription factor")


## Create a dataframe from annotation data

annotation_df <- as.data.frame(peakAnno)

annotation.to.exclude <- c("Distal Intergenic", "Downstream (1-2kb)", 
                           "Downstream (<1kb)", "Downstream (2-3kb)")

targets <- subset(annotation_df, !(annotation %in% annotation.to.exclude))
regulome <- targets$geneId

print("", quote = F)
print(paste0(c(length(regulome), "genes have been determined to be a target for", args[2], ". They are now being stored in a txt file."),collapse = " "), quote = F)
print("", quote = F)

## Write a txt that stores the target genes (i.e. the regulome)

write(x = regulome,file = paste0(c(args[2],"_target_genes.txt"),collapse = ""))

## Get all genes from arabidopsis database

genes_arabidopsis <- as.data.frame(genes(txdb))
genes_arabidopsis_names <- rownames(genes_arabidopsis)

## GO terms enrichment

print("", quote = F)
print("GO Enrichment analysis is taking place. Given a vector of genes like the one that was previously generated, enrichGO (from clusterProfiler package) will return the enrichment GO categories after False Discovery Rate (FDR) control.", quote = F)
print("", quote = F)

ego <- enrichGO(gene = regulome, universe = genes_arabidopsis_names, OrgDb = annotation_atha, ont = "ALL", pAdjustMethod = "BH", pvalueCutoff  = 0.01, qvalueCutoff  = 0.05, readable = TRUE, keyType = "TAIR")

head(ego)

## GO terms enrichment visualization

barplot(ego, showCategory=15,main="GO Enrichment Analysis")
dotplot(ego, showCategory=20)

## Cluster-profiling of GO terms

## Creating a network is possible with cnetplot, but it requires a specific format that can be achieved
## with setReadable (available at DOSE package)

edox <- setReadable(ego, 'org.At.tair.db', 'TAIR')

cnetplot(x = edox, colorEdge = TRUE)

print("==========================", quote = F)
print("| R script has finished! |", quote = F)
print("==========================", quote = F)
